packages/polymath-js/src/contracts/ManualApprovalTransferManager.js

```diff
@@ -63,20 +63,40 @@ export default class ManualApprovalTransferManager extends Contract {
     for (let i = 0; i < approvals['0'].length; i++) {
       let a = {};
 
-      a.id = approvals['0'][i] + approvals['1'][i];
+      a.id = (approvals['0'][i] + approvals['1'][i]).toLowerCase();
       a.fromAddress = approvals['0'][i];
       a.toAddress = approvals['1'][i];
-      let events = await this._contractWS.getPastEvents('AddManualApproval', {
+      a.expiry = approvals['4'][i];
+
+      const events = await this._contractWS.getPastEvents('AddManualApproval', {
         filter: { _from: a.fromAddress, _to: a.toAddress },
         fromBlock: 0,
         toBlock: 'latest',
       });
-      a.txHash = events[events.length - 1].transactionHash;
+
+      const modifyEvents = await this._contractWS.getPastEvents(
+        'ModifyManualApproval',
+        {
+          filter: { _from: a.fromAddress, _to: a.toAddress },
+          fromBlock: 0,
+          toBlock: 'latest',
+        }
+      );
+
+      if (
+        modifyEvents.length > 0 &&
+        modifyEvents[modifyEvents.length - 1].blockNumber >
```

Comment:
What's the purpose of this condition? Is it to get txHash of the modification event (if any), otherwise get txHash of  the approval event? Are we sure that events are sorted by block number, and ascendingly? Otherwise, the last event of in the array might not be the event we're looking for.

Why are we storing txHash at the first place?

I would merge both events arrays, sort them by block number, and then select last element of that merged array.

Reply:
The last value in the array is always the latest event from the getPastEvents result. The hash needs to be stored because it is clickable on the table to get to etherscan. Merging both arrays is not necessary since you will have excess data that you do not need to go through. You only need the last of the two events.